matching-core/src/main/java/com/graphhopper/matching/GPXExtension.java

```diff
@@ -17,36 +17,95 @@
  */
 package com.graphhopper.matching;
 
+import com.graphhopper.routing.VirtualEdgeIteratorState;
 import com.graphhopper.storage.index.QueryResult;
+import com.graphhopper.util.EdgeIteratorState;
 import com.graphhopper.util.GPXEntry;
 
 /**
+ * During map matching this class represents a map matching candidate, i.e. a potential snapped
+ * point of a GPX entry. After map matching, this class represents the map matched point of
+ * an GPX entry.
+ *
+ * An instance of this class represents either an undirected real node or a directed virtual node.
+ * If this represents a directed virtual node then incoming paths from any previous GPXExtension
+ * should arrive through {@link #getIncomingVirtualEdge()} and outgoing paths to any following
+ * GPXExtension should start with {@link #getOutgoingVirtualEdge()}. This is achieved by
+ * penalizing other edges for routing. Note that virtual nodes are always connected to their
+ * adjacent nodes via 2 virtual edges (not counting reverse virtual edges).
  *
  * @author Peter Karich
+ * @author kodonnell
+ * @author Stefan Holder
  */
 public class GPXExtension {
-    final GPXEntry entry;
-    final QueryResult queryResult;
-    final int gpxListIndex;
+    private final GPXEntry entry;
+    private final QueryResult queryResult;
+    private final boolean isDirected;
+    private final EdgeIteratorState incomingVirtualEdge;
+    private final EdgeIteratorState outgoingVirtualEdge;
 
-    public GPXExtension(GPXEntry entry, QueryResult queryResult, int gpxListIndex) {
+    /**
+     * Creates an undirected candidate for a real node.
+     */
+    public GPXExtension(GPXEntry entry, QueryResult queryResult) {
         this.entry = entry;
         this.queryResult = queryResult;
-        this.gpxListIndex = gpxListIndex;
+        this.isDirected = false;
+        this.incomingVirtualEdge = null;
+        this.outgoingVirtualEdge = null;
     }
 
-    @Override
-    public String toString() {
-        return "entry:" + entry
-                + ", query distance:" + queryResult.getQueryDistance()
-                + ", gpxListIndex:" + gpxListIndex;
+    /**
+     * Creates a directed candidate for a virtual node.
+     */
+    public GPXExtension(GPXEntry entry, QueryResult queryResult,
+                        VirtualEdgeIteratorState incomingVirtualEdge,
+                        VirtualEdgeIteratorState outgoingVirtualEdge) {
+        this.entry = entry;
+        this.queryResult = queryResult;
+        this.isDirected = true;
+        this.incomingVirtualEdge = incomingVirtualEdge;
+        this.outgoingVirtualEdge = outgoingVirtualEdge;
+    }
+
+    public GPXEntry getEntry() {
+        return entry;
     }
 
     public QueryResult getQueryResult() {
-        return this.queryResult;
+        return queryResult;
     }
 
-    public GPXEntry getEntry() {
-        return entry;
+    /**
+     * Returns true if the snapped point is a virtual node, otherwise returns false.
```

Comment:
I see your point but from a map matching point of view, we either have directed or undirected candidates. Currently, all virtual nodes are transformed into directed candidates but we might later have directed candidates also for real nodes later as discussed before. So the method name `isDirected` would be more stable than `isVirtual`. Also the javadoc of this method is meant to be read in combination with the [class javadoc](https://github.com/graphhopper/map-matching/pull/88/files#diff-eb87539fc299ca85661a06916b4511f6R30), which should explain the whole picture.